File: Codeforces/div810/C.py

def inp():
    return list(map(int, input().split()))


for _ in range(int(input())):
    m, n, k = inp()
    a = inp()
    a2 = list(a)
    if sum(a) < m * n:
        print('NO')
        continue
    ok = False
    ki = 0
    ni = 0
    pos1 = False
    pos2 = False
    if n % 2:
        for i in range(k):
            add = a[i] // m
            if add >= 3:
                ni += 3
                pos1 = True
                a[i] -= m * 3
                break
    else:
        pos1 = True
    if pos1:
        while True:
            if ki >= k:
                break
            add = a[ki] // m
            a[ki] -= add * m
            if add > 1:
                ni += add
            else:
                ki += 1
            if ni >= n:
                ok = True
                break
    if ok:
        print('YES')
        continue

    a = a2
    ki = 0
    del ni
    mi = 0
    if m % 2:
        for i in range(k):
            add = a[i] // n
            if add >= 3:
                mi += 3
                pos2 = True
                a[i] -= n * 3
                break
    else:
        pos2 = True
    if pos2:
        while True:
            if ki >= k:
                break
            add = a[ki] // n
            a[ki] -= add * n
            if add > 1:
                mi += add
            else:
                ki += 1
            if mi >= m:
                ok = True
                break

    if ok:
        print('YES')
    else:
        print('NO')

File: Codeforces/div810/C.cpp

#include <iostream>
#include <vector>

using namespace::std;

int main()
{
	int t;
	cin >> t;
	while (t--)
	{
		int m, n, k;
		cin >> m >> n >> k;
		vector<int> a(k);
		for (int i=0; i < k; ++i)
			cin >> a.at(i);
		vector<int> a2(a);
		long long sm = 0;
		for (int i : a)
			sm += i;
		if (sm < m * n)
		{
			cout << "NO\n";
			continue;
		}
		bool ok = false;
		int ki = 0;
		int ni = 0;
		bool pos1 = false;
		bool pos2 = false;
		if (n % 2)
		{
			for (int i=0; i < k; ++i)
			{
				int add = a.at(i) / m;
				if (add >= 3)
				{
					ni += 3;
					pos1 = true;
					a.at(i) -= m * 3;
					break ;
				}
			}
		}
		else
		{
			pos1 = true;
		}
		if (pos1)
		{
			while (true)
			{
				if (ki >= k)
					break;
				int add = a.at(ki) / m;
				a.at(ki) -= add * m;
				if (add > 1)
					ni += add;
				else
					++ki;
				if (ni >= n)
				{
					ok = true;
					break ;
				}
			}
		}
		if (ok)
		{
			cout << "YES\n";
			continue ;
		}
		a.swap(a2);
		ki = 0;
		int mi = 0;
		if (m % 2)
		{
			for (int i=0; i < k; ++i)
			{
				int add = a.at(i) / n;
				if (add >= 3)
				{
					mi += 3;
					pos2 = true;
					a.at(i) -= n * 3;
					break ;
				}
			}
		}
		else
			pos2 = true;
		if (pos2)
		{
			while (true)
			{
				if (ki >= k)
					break ;
				int add = a.at(ki) / n;
				a.at(ki) -= add * n;
				if (add > 1)
					mi += add;
				else
					++ki;
				if (mi >= m)
				{
					ok = true;
					break ;
				}
			}
		}
		if (ok)
			cout << "YES\n";
		else
			cout << "NO\n";
	}
}

